ggml : fix padding in timestep embedding kernels

danbev opened the pull request:
This commit applies the same changes as were applied in #15917 to the cuda, metal, opencl, sycl, and vulkan backends.

Refs: https://github.com/ggml-org/llama.cpp/pull/15917

danbev committed:
ggml-cuda : fix padding in timestep embedding kernel

This commit removes the zeroing out of the last dimension now that we
are not adding the extra padding dimension.

danbev committed:
ggml-sycl : fix padding in timestep embedding kernel

This commit fixes the zero padding for odd dimensions in
the timestep embedding kernel.

danbev committed:
ggml-opencl : fix padding in timestep embedding kernel

This commit fixes the zero padding for odd dimensions in
the timestep embedding kernel.

danbev committed:
ggml-metal : fix padding in timestep embedding kernel

This commit fixes the zero padding for odd dimensions in
the timestep embedding kernel

danbev committed:
ggml-vulkan : fix padding in timestep embedding kernel

This commit fixes the zero padding for odd dimensions in
the timestep embedding kernel.

ggerganov reviewed (approved):
I think we should add a comment to the declaration of this function to clarify that it will pad the tensor if it has odd number of dimensions. So here:

https://github.com/ggml-org/llama.cpp/blob/28b5f190ef1dbea5edf82dbc8b4407b721fadd13/ggml/include/ggml.h#L2123-L2130

Change to:

```c
// return: [N, PAD(dim, 2)]
```

Btw, not sure what was the reasoning for this padding logic. It does not seem to be present in the reference Python implementation. Maybe @leejet can clarify?

ggerganov commented:
@danbev I think the problem is that for input `dim=15` the Python code produces output dimension of `15`, while `ggml` produces `16`. At least this is what I understand from reading the implementation in `ggml.c`:

https://github.com/ggml-org/llama.cpp/blob/261e6a20ffdb79c4875e674b4f6b514bc73cff8f/ggml/src/ggml.c#L4919-L4932

danbev committed:
ggml : remove adding extra dim timestep embedding

This commit updates the ggml_timestep_embedding function to no longer
add an extra dimension when the specified dimension is odd.

The motivation for this change is that this introduces an unnecessary
dimension when the dimension is odd, which caused an issue in the
kernels which were not expecting this extra dimension and it resulted in
uninitialized memory for the second to last dimension.

danbev committed:
ggml-cpu : fix padding in timestep embedding function

This commit removes the zeroing out of the last dimension now that we
are not adding the extra padding dimension.
